File: commands/userAdd.js
const chalk = require('chalk');
const {User} = require('../models/User.js');
const {request} = require('../libraries/request.js');
const {prompt,confirm} = require('../libraries/prompt.js');

module.exports = {
  async userAdd(name){
    const newUser = new User({name});

    let answer;

    do{
      answer = await confirm(chalk.green(`Is this correct, create user: "${newUser.name}"?`));
      if(!answer){
        console.log(chalk.green('No problem, let\'s try again.'));
        newUser.name = await prompt(chalk.green('Please enter name: '));
      } //end if
    }while(!answer)
    do{
      newUser.email = await prompt(chalk.green('Please enter email: '));
      answer = await confirm(chalk.green(`Is this correct: "${newUser.email}"?`));
      if(!answer) console.log(chalk.green('No problem, let\'s try again.'));
    }while(!answer)
    newUser.publicKey = await prompt(chalk.green('Please enter public key: '));
    newUser.permissions.viewUsers = await confirm(chalk.green('Allow view all users permission? '));
    newUser.permissions.editUsers = await confirm(chalk.green('Allow edit all users permission? '));
    newUser.permissions.viewScopeNames = await confirm(chalk.green('Allow view all scopes permission? '));
    newUser.permissions.createScopes = await confirm(chalk.green('Allow create scopes permission? '));
    do{
      answer = await prompt(chalk.green('Enter scopes to access separated by commas: '));
      const scopeNames = answer.split(',').map(n=> n.trim());

      for(let i=0;i<scopeNames.length;i++){
        answer = await confirm(chalk.green(`Allow full access (Y) or view access to scope "${scopeNames[i]}"(n)?`));
        newUser.permissions.scopes.push({name: scopeNames[i],value: answer?'edit':'view'});
      } //end for
      answer = await confirm(chalk.green(`Is this correct: "${
        newUser.permissions.scopes.map(s=> `(${s.name}): ${s.value}`).join()
      }"?`));
      if(!answer){
        newUser.permissions.scopes.length = 0;
        console.log(chalk.green('No problem, let\'s try again.'));
      } //end if
    }while(!answer)
    try{
      const res = await request('userAdd',newUser);

      if(res.success){
        console.log(chalk.green(`User "${newUser.name}" added successfully!`));
      }else{
        console.log(chalk.red(res.error));
      } //end if
    }catch(err){
      console.log(chalk.red('Problem connecting to server.'));
      console.log(chalk.red(err));
    }
  }
};

File: cmd/user_add.go
package cmd

import (
	"encoding/json"
	"fmt"

	"github.com/fatih/color"
	"github.com/nate/credential-broker/client"
	"github.com/nate/credential-broker/model"
	"github.com/nate/credential-broker/prompt"
	"github.com/spf13/cobra"
)

var userAddCmd = &cobra.Command{
	Use:   "add [name]",
	Short: "Add a new user",
	Args:  cobra.ExactArgs(1),
	RunE:  runUserAdd,
}

func runUserAdd(cmd *cobra.Command, args []string) error {
	name := args[0]

	email, err := prompt.Text("Email")
	if err != nil {
		return err
	}

	// Read the new user's public key
	keyPath, err := prompt.Text("Path to user's public key file")
	if err != nil {
		return err
	}
	keyBytes, err := readFileBytes(keyPath)
	if err != nil {
		return fmt.Errorf("failed to read public key: %w", err)
	}

	fmt.Println("Set permissions:")
	viewUsers, _ := prompt.Confirm("  View users?")
	editUsers, _ := prompt.Confirm("  Edit users?")
	viewScopeNames, _ := prompt.Confirm("  View scope names?")
	createScopes, _ := prompt.Confirm("  Create scopes?")

	body := map[string]interface{}{
		"name":  name,
		"email": email,
		"key":   string(keyBytes),
		"permissions": model.Permissions{
			ViewUsers:      viewUsers,
			EditUsers:      editUsers,
			ViewScopeNames: viewScopeNames,
			CreateScopes:   createScopes,
			Scopes:         []model.ScopePermission{},
		},
	}

	bodyJSON, err := json.Marshal(body)
	if err != nil {
		return err
	}

	resp, err := client.Request("userAdd", string(bodyJSON))
	if err != nil {
		return err
	}

	var result map[string]string
	if err := json.Unmarshal([]byte(resp), &result); err != nil {
		color.Green("%s", resp)
		return nil
	}

	if msg, ok := result["success"]; ok {
		color.Green("%s", msg)
	} else if msg, ok := result["error"]; ok {
		color.Red("%s", msg)
	}

	return nil
}

func readFileBytes(path string) ([]byte, error) {
	return readFileContent(path)
}
